src/navigation/gate.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+from typing import ClassVar, Optional
+from unicodedata import normalize
+from context import Gate
+
+import numpy as np
+
+from context import Context, Environment
+from state import BaseState
+from trajectory import Trajectory
+from dataclasses import dataclass
+from drive import get_drive_command
+from util.np_utils import normalized, perpendicular_2d
+
+STOP_THRESH = 0.2
+DRIVE_FWD_THRESH = 0.95
+
+APPROACH_DISTANCE = 2.0
+
+
+@dataclass
+class GateTrajectory(Trajectory):
+    @classmethod
+    def spider_gate_trajectory(cls, approach_distance: float, gate: Gate, rover_position: np.ndarray) -> GateTrajectory:
+        """
+        Generates the "Spider" path through the gate
+        :param approach_distance: distance to the point straight out from the gate that the rover will drive to
+        :param gate:    Gate object representing the gate that the rover will drive through
+        :param rover_position: position vector of the rover
+        :return:            GateTrajectory object containing the coordinates the rover will need to traverse
+        """
+
+        # first we get the positions of the two posts
+        post1 = gate.post1
+        post2 = gate.post2
+
+        center = (post1 + post2) / 2
+        # the direction of the post is just the normalized vector from post1 to post2
+        post_direction = normalized(post2 - post1)
+        perpendicular = perpendicular_2d(post_direction)
+
+        # appraoch points are the points that are directly out from the center (a straight line) of
+        # the gate "approach_distance" away
+        possible_approach_points = [
+            approach_distance * perpendicular + center,
+            -approach_distance * perpendicular + center,
+        ]
+
+        # prep points are points that are direclty out in either direction from the posts
+        # the idea here is that if we go to the closest prep point then an approach point,
+        # we will never collide with the post
+        possible_preparation_points = [
+            (2 * approach_distance * perpendicular) + post1,
+            (2 * approach_distance * perpendicular) + post2,
+            (-2 * approach_distance * perpendicular) + post1,
+            (-2 * approach_distance * perpendicular) + post2,
+        ]
+
+        # get closest prepration point
+        prep_distance_to_rover = [np.linalg.norm(point - rover_position[0:2]) for point in possible_preparation_points]
```

Comment:
use braodcasting rules instead of list comprehension

Reply:
should be addressed